agent/engine/docker_task_engine.go

```diff
@@ -347,34 +355,38 @@ func (engine *DockerTaskEngine) wakeUpTaskQueueMonitor() {
 	}
 }
 
-func (engine *DockerTaskEngine) topTask() (*managedTask, error) {
+// getTaskByIndex returns the task from waitingTaskQueue at the given index
+func (engine *DockerTaskEngine) getTaskByIndex(index int) (*managedTask, error) {
 	engine.waitingTasksLock.Lock()
 	defer engine.waitingTasksLock.Unlock()
-	if len(engine.waitingTaskQueue) > 0 {
-		return engine.waitingTaskQueue[0], nil
+	waitingTaskQueueLength := len(engine.waitingTaskQueue)
+	// empty queue
+	if waitingTaskQueueLength == 0 {
+		return nil, fmt.Errorf("no tasks in the waiting queue")
 	}
-	return nil, fmt.Errorf("no tasks in waiting queue")
+	// invalid index
+	if index < 0 || index >= waitingTaskQueueLength {
+		return nil, fmt.Errorf("invalid index: %v", index)
+	}
+	return engine.waitingTaskQueue[index], nil
 }
 
 func (engine *DockerTaskEngine) enqueueTask(task *managedTask) {
 	engine.waitingTasksLock.Lock()
 	engine.waitingTaskQueue = append(engine.waitingTaskQueue, task)
 	engine.waitingTasksLock.Unlock()
-	logger.Debug("Enqueued task in Waiting Task Queue", logger.Fields{field.TaskARN: task.Arn})
+	logger.Debug("Enqueued task in the waiting task queue", logger.Fields{field.TaskARN: task.Arn})
 	engine.wakeUpTaskQueueMonitor()
 }
 
-func (engine *DockerTaskEngine) dequeueTask() (*managedTask, error) {
+// dequeueTaskByIndex dequeues a task from the waitingTaskQueue at the given index
+func (engine *DockerTaskEngine) dequeueTaskByIndex(index int) *managedTask {
 	engine.waitingTasksLock.Lock()
 	defer engine.waitingTasksLock.Unlock()
-	if len(engine.waitingTaskQueue) > 0 {
-		task := engine.waitingTaskQueue[0]
-		engine.waitingTaskQueue = engine.waitingTaskQueue[1:]
-		logger.Debug("Dequeued task from Waiting Task Queue", logger.Fields{field.TaskARN: task.Arn})
-		return task, nil
-	}
-
-	return nil, fmt.Errorf("no tasks in waiting queue")
+	task := engine.waitingTaskQueue[index]
```

Comment:
Q: should we validate the index here before dequeuing?

Reply:
good callout, I went back and forth in my head about this. The assumption here is the current callers call this method with valid indexes. I will update this to remove this assumption and validate to err on the safe side.

Reply:
updated

Reply:
thanks!